Actually log in newly created clients

fortysixandtwo opened the pull request:
See individual commits.

This allows newly created accounts to log in.

However, they will currently not be properly saved into the keyring
and will also create a new device id for each log in

fortysixandtwo committed:
matrix: Enable CmClient

Otherwise a completely new account won't ever log in.

fortysixandtwo committed:
matrix: Demote warning about creating a new client to debug level

The latest libcmatrix will not blindly load all accounts.

However, the crash might still happen if there are multiple accounts
set up for eigenvalue
(libcmatrix checks the keyring for secrets
with a label that contains the app id)
as we currently only set the sync callback for a single client.

In any case creating a new client should not warrant raising a warning.

fortysixandtwo committed:
matrix: Disable other clients

Whether a client is enabled or not is stored in the keyring.
libcmatrix will try connecting enabled clients
(even if it's not the one we set up in our account config)
which likely leads to a crash, as we've never set their sync callback.

fortysixandtwo committed:
matrix: Save new client into DB and keyring

Needs new API API from the updated libcmatrix fork, namely
- cm_matrix_save_client_sync()
- cm_utils_get_homeserver_sync()

The homeserver and login ID need to be properly set before we can
successfully save the client.
cm_utils_get_homeserver_sync() is needed because we cannot simply
use the server part of "@user:example.com" as server deligation may be
in play and we need to query .well-known/matrix/client to identify the
homeserver.

fortysixandtwo commented:
Without 40932e493ed41eeb2199596a238c3fee3ff01eaa we log in just fine, but we don't save the client to the db and keyring and we see a couple of errors in the logs:

```
cm-room-LOG-0x100: 13:14:08.463: (0x5555559cca80) New timeline events count: 20
(eigenvalue:518939): cm-room-WARNING **: 13:14:08.463: (0x5555559cca80) Saving room details error: Error getting account id
(eigenvalue:518939): cm-room-WARNING **: 13:14:08.463: (0x5555559cca80) Saving room details error: Error getting account id
(eigenvalue:518939): cm-db-DEBUG: 13:14:08.463: Error getting session: Account or Room not found in db
[...]
(eigenvalue:518939): cm-room-WARNING **: 13:14:08.464: (0x555555cfc8d0) Saving room details error: Error getting account id
(eigenvalue:518939): cm-db-DEBUG: 13:14:08.464: Error getting session: Error getting account id
```

With 40932e493ed41eeb2199596a238c3fee3ff01eaa the client gets saved to the db and keyring, but it won't be able to login (with the rest of the code as is). I've narrowed it down to `cm_client_set_homeserver ()`

I love sideeffects.

fortysixandtwo commented:
ok, so chatty does this on account creation:

```
  cm_client_set_homeserver (cm_client, uri_has_prefix ? uri : uri_prefixed);
  cm_client_set_password (cm_client, password);
  cm_client_get_homeserver_async (cm_client, self->cancellable,
                                  matrix_home_server_verify_cb,
                                  self);
```

and from  `matrix_home_server_verify_cb()` the saving related call chain looks like `chatty_manager_save_account_async (chatty_manager_get_default (), CHATTY_ACCOUNT (account), NULL, settings_save_account_cb, self);` -> `chatty_matrix_save_account_async (self->matrix, account, cancellable, manager_save_account_cb,  g_steal_pointer (&task));`

so it does one `set_homeserver()` `get_homeserver_async()` round trip before saving. I'll give that a try

fortysixandtwo commented:
Ok, I see now what the problem was:
my homeserver is https://matrix.fortysixandtwo.eu *not* https://fortysixandtwo.eu

In the regular flow (when we don't set the homeserver ourselves) the homeserver URI get's grabbed from the .well-known.

I'll check what happens if I simply relax libcmatrix to not require the homeserver set when trying to save the client.

Edit: Probably that's not a great idea as the homeserver is one of the attributes of our `SecretSchema`

agx commented:
> This allows newly created accounts to log in.

Great! This would also allow us to get rid of the config file.

> However, they will currently not be properly saved into the keyring

I do get accounts saved in the keyring:

    org.sigxcpu.Eigenvalue Matrix password for "@asdf3:sigxcpu.org"

even without that code but I assume this is likely just the account migration code in chatty triggering.

fortysixandtwo commented:
> Great! This would also allow us to get rid of the config file.

So in that case we'd just get prompted for the username and pw on the CLI?

And in case of preexisting account get presented with a list of accounts and allow a user to enter which account to login to, e.g.
```
[0] @account1:server1
[1] @account2:server2

[c] Create new account

> Please enter account to login to or 'c' to create a new account
```

fortysixandtwo commented:
> I do get accounts saved in the keyring:
> 
> ```
> org.sigxcpu.Eigenvalue Matrix password for "@asdf3:sigxcpu.org"
> ```
> 
> even without that code but I assume this is likely just the account migration code in chatty triggering.

I would assume so. Did you try with https://source.puri.sm/Librem5/libcmatrix/-/merge_requests/87 ?
(I'll bump the commit in the libcmatrix subproject once that MR has landed)

fortysixandtwo committed:
matrix: Only set password, device name and login id for new clients

This information get's saved into the DB and keyring anyway,
so we don't need to set it each time.

fortysixandtwo commented:
v2: 
- needs the **current** https://source.puri.sm/Librem5/libcmatrix/-/merge_requests/87
-  fetch the homeserver via new API
-  only set password, device name and login id for newly created devices

How I tested this PR:
- remove ~/.local/share/eigenvalue/matrix.db
- remove any org.sigxcpu.eigenvalue entry from the keyring
- run eigenvalue and marvel at a successful login with no warnings/criticals
- quit
- start again and still marvel at a successful login with no warnings/criticals

agx commented:
> So in that case we'd just get prompted for the username and pw on the CLI?

I'd say it should be similar to rooms:

`/accounts`: lists the known accounts 
`/add-account <matrixid> <password>` : add account
`/account-details <matrixid>`: Show details about an account

(this doesn't have to happen in this MR, just wanted mention it in case it helps laying out the code).

fortysixandtwo commented:
> > So in that case we'd just get prompted for the username and pw on the CLI?
> 
> I'd say it should be similar to rooms:
> 
> `/accounts`: lists the known accounts `/add-account <matrixid> <password>` : add account `/account-details <matrixid>`: Show details about an account

that seems more than sensible!

> (this doesn't have to happen in this MR, just wanted mention it in case it helps laying out the code).

does not make a difference for this MR I think and I'm not tackling this one right now (want to work on joining rooms next)

agx commented on src/ev-matrix.c:
```diff
@@ -175,33 +175,43 @@ on_matrix_open (GObject *object, GAsyncResult *result, gpointer user_data)
       client = g_steal_pointer (&c);
-      break;
+    } else {
+      cm_client_set_enabled (c, FALSE);
```
```suggestion
      /* FIXME: libcmatrix shouldn't enable clients by default */
      cm_client_set_enabled (c, FALSE);
```

fortysixandtwo replied:
I don't think libcmatrix does (but I may be wrong here). Otherwise I wouldn't have had to explicitly `cm_client_set_enabled (client, TRUE)`  below for the newly created account to login.

Instead this just ensures that we're only ever enabling a single one.

fortysixandtwo replied:
I've not included this change, but reworded the commit message to make the current situation hopefully clearer.

agx replied:
But we break out of the loop, how would we enable multiple ones?

fortysixandtwo replied:
> But we break out of the loop

Not anymore? `break` has been removed

> how would we enable multiple ones?

by enabling the ones we want later on (as we do now for the single client/account).

I briefly considered multiple accounts, but since the code in the beginning only looks for a single account `matrix-00` I thought this would be a problem for the future.

agx replied:
Again myself explaining things badly, sorry:

> Not anymore? break has been removed

Yes, my question is: why not just use the `break` as before. That would only ever enable a single account.

>  how would we enable multiple ones?

(that was related to: if we `break` out of the loop how can we ever end up enabling an account accidentally). I am basically trying to understand how you think we could get into a situation that requires `cm_client_set_enabled (c, FALSE);` if we `break` out of the loop after enabling the first account?

fortysixandtwo replied:
I've not checked in detail, but I think`cm_client_set_enabled ()` persists as it is part of the stored secret.

So if we switch the account (by changing the username/pw of `matrix-00` in our config) I would assume (untested as of yet) that the old account would stay enabled which should then lead to the failed assertion of a sync callback not set.

Instead of the theory crafting it would probably be wise to confirm that this is actually what would happen,
but this will have to wait until tomorrow.

agx replied:
Which circles us back to my initial whish for a `FIXME`: libcmatrix shouldn't enable clients by default :smile: - it's fine if it keeps the `enabled` state tracked in the secret store but it shouldn't start *enabling* when creating a matrix client object (blowing up if there's no callback connected, etc). That should be an explicit operation like:

- create `CmMatrix`
- get a list of clients for this application / configured accounts
- allow the app to check which should be enabled and then call `set_enabled` on them to actually start processing

For the simple use case we could add a `auto-enable` property to `CmMatrix` to keep the current behavior but I'd argue that's rarely what an app wants.

So from a code / commit message point of view this looks fine to me. It would just be good to add the `FIXME`.

agx commented:
Thanks for looking into this. Having this cleaned up will make other clients way more simple.

As we're using a subproject: Instead of linking to MRs in commit messages you can just point the subproject to a sensible branch. We'll update that to `main` once the corresponding MR is merged. 

This also ensures ppl are testing exactly that combination of eigenvalue and libcmatrix.

fortysixandtwo committed:
libcmatrix: Switch to main branch

It contains the new synchronous API for saving clients.

fortysixandtwo commented:
v3:
- Update libcmatrix subproject to include necessary API
-  Reworded commit messages (including, but not limited to **not** linking to the libcmatrix MRs)

agx commented:
> Update libcmatrix subproject to include necessary API

This (at revision e069131) fails for me (when I drop `-Dtests=false` from `meson.build` to make `meson setup`) pass like

```
../src/ev-matrix.c: In function ‘on_matrix_open’:
../src/ev-matrix.c:208:54: error: passing argument 3 of ‘cm_matrix_save_client_sync’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  208 |     if (!cm_matrix_save_client_sync (matrix, client, &error))
      |                                                      ^~~~~~
      |                                                      |
      |                                                      GError ** {aka struct _GError **}
In file included from ../subprojects/libcmatrix/src/cmatrix.h:29,
                 from ../src/ev-matrix.c:18:
../subprojects/libcmatrix/src/cm-matrix.h:50:59: note: expected ‘GCancellable *’ {aka ‘struct _GCancellable *’} but argument is of type ‘GError **’ {aka ‘struct _GError **’}
   50 |                                             GCancellable *cancellable,
      |                                             ~~~~~~~~~~~~~~^~~~~~~~~~~
../src/ev-matrix.c:208:10: error: too few arguments to function ‘cm_matrix_save_client_sync’
  208 |     if (!cm_matrix_save_client_sync (matrix, client, &error))
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
../subprojects/libcmatrix/src/cm-matrix.h:48:13: note: declared here
   48 | gboolean    cm_matrix_save_client_sync     (CmMatrix *self,
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

agx commented:
Likely due to https://source.puri.sm/Librem5/libcmatrix/-/merge_requests/87#note_273745 . I'll wait for an update then :smile:

fortysixandtwo commented:
You're too fast :rocket:  I only just updated the libcmatrix side ;)

fortysixandtwo commented:
v4: updated to adjust to API breakage

agx commented:
Builds now, thanks. I'm seeing  a new crash on logout now when the sync for each client is stopped:

```
#0  g_type_check_instance_is_fundamentally_a (type_instance=type_instance@entry=0x562466994ae0, fundamental_type=fundamental_type@entry=0x50 [GObject])
    at ../../../gobject/gtype.c:4153
#1  0x00007f9ae34b4ab6 in g_object_ref (_object=0x562466994ae0) at ../../../gobject/gobject.c:4238
#2  0x00005624661dc47a in matrix_stop (self=0x56246694e180 [CmMatrix]) at ../subprojects/libcmatrix/src/cm-matrix.c:192
#3  cm_matrix_finalize (object=0x56246694e180 [CmMatrix]) at ../subprojects/libcmatrix/src/cm-matrix.c:224
#4  0x00007f9ae34b52c3 in g_object_unref (_object=0x56246694e180) at ../../../gobject/gobject.c:4484
#5  0x00005624661c0904 in ev_matrix_destroy () at ../src/ev-matrix.c:242
#6  0x00005624661be312 in ev_application_shutdown (app=0x5624669369a0 [EvApplication]) at ../src/ev-application.c:78
#10 0x00007f9ae34cafc3 in <emit signal 'shutdown' on instance 0x5624669369a0 [EvApplication]>
    (instance=instance@entry=0x5624669369a0, signal_id=<optimized out>, detail=detail@entry=0) at ../../../gobject/gsignal.c:3583
    #7  0x00007f9ae34af939 in _g_closure_invoke_va (closure=0x562466935d40, return_value=0x0, instance=0x5624669369a0, args=0x7fffb6242200, n_params=0, param_types=0x0)
    at ../../../gobject/gclosure.c:897
    #8  0x00007f9ae34c533f in signal_emit_valist_unlocked
    (instance=instance@entry=0x5624669369a0, signal_id=signal_id@entry=7, detail=detail@entry=0, var_args=var_args@entry=0x7fffb6242200) at ../../../gobject/gsignal.c:3424
    #9  0x00007f9ae34caf06 in g_signal_emit_valist (instance=0x5624669369a0, signal_id=7, detail=0, var_args=0x7fffb6242200) at ../../../gobject/gsignal.c:3263
#11 0x00007f9ae35e1722 in g_application_run (application=application@entry=0x5624669369a0 [EvApplication], argc=argc@entry=1, argv=argv@entry=0x7fffb6242468)
    at ../../../gio/gapplication.c:2718
#12 0x00005624661be012 in main (argc=1, argv=0x7fffb6242468) at ../src/main.c:42
```

agx commented:
There's also new criticals since 2nd startup:

```
_build/run 
+ exec /var/scratch/librem5/eigenvalue/eigenvalue/_build/src/eigenvalue
Ev> 
(eigenvalue:406724): GLib-GObject-CRITICAL **: 20:42:00.902: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

(eigenvalue:406724): cm-client-CRITICAL **: 20:42:00.902: cm_client_get_account: assertion 'CM_IS_CLIENT (self)' failed

(eigenvalue:406724): cm-account-CRITICAL **: 20:42:00.902: cm_account_get_login_id: assertion 'CM_IS_ACCOUNT (self)' failed

(eigenvalue:406724): cm-client-CRITICAL **: 20:42:00.902: cm_client_set_enabled: assertion 'CM_IS_CLIENT (self)' failed
Logging in @asdf3:sigxcpu.org
cm-room-LOG-0x100: 20:42:00.902: (0x7efc7c02d0e0) new room '!hK###############K:s#####u.o#g' from json
```

With the first one being

```
(gdb) bt
#0  g_logv (log_domain=0x7fd962256f93 "GLib-GObject", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7ffcca978f00)
    at ../../../glib/gmessages.c:1277
#1  0x00007fd962129aa3 in g_log
    (log_domain=log_domain@entry=0x7fd962256f93 "GLib-GObject", log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7fd962182397 "%s: assertion '%s' failed") at ../../../glib/gmessages.c:1315
#2  0x00007fd96212aaad in g_return_if_fail_warning
    (log_domain=log_domain@entry=0x7fd962256f93 "GLib-GObject", pretty_function=pretty_function@entry=0x7fd96225f918 <__func__.28> "g_object_ref", expression=expression@entry=0x7fd9622578aa "G_IS_OBJECT (object)") at ../../../glib/gmessages.c:3000
#3  0x00007fd96222cb1c in g_object_ref (_object=<optimized out>) at ../../../gobject/gobject.c:4238
#4  g_object_ref (_object=<optimized out>) at ../../../gobject/gobject.c:4232
#5  0x000055d410ca7e34 in on_matrix_open (object=<optimized out>, result=<optimized out>, user_data=<optimized out>) at ../src/ev-matrix.c:171
#6  0x00007fd962329733 in g_task_return_now (task=task@entry=0x55d4139ac2e0 [GTask]) at ../../../gio/gtask.c:1361
#7  0x00007fd96232a3d3 in g_task_return (type=<optimized out>, task=0x55d4139ac2e0 [GTask]) at ../../../gio/gtask.c:1430
#8  g_task_return (task=0x55d4139ac2e0 [GTask], type=<optimized out>) at ../../../gio/gtask.c:1387
#9  0x000055d410cc5e5c in db_open_cb (obj=<optimized out>, result=<optimized out>, user_data=0x55d4139ac2e0) at ../subprojects/libcmatrix/src/cm-matrix.c:453
#10 0x00007fd962329733 in g_task_return_now (task=task@entry=0x55d4139d9680 [GTask]) at ../../../gio/gtask.c:1361
#11 0x00007fd96232976d in complete_in_idle_cb (task=0x55d4139d9680) at ../../../gio/gtask.c:1375
#12 0x00007fd96211ee2f in g_main_dispatch (context=context@entry=0x55d413982a80) at ../../../glib/gmain.c:3344
#13 0x00007fd962120eb7 in g_main_context_dispatch_unlocked (context=0x55d413982a80) at ../../../glib/gmain.c:4152
#14 g_main_context_iterate_unlocked (context=context@entry=0x55d413982a80, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>)
    at ../../../glib/gmain.c:4217
#15 0x00007fd9621214d0 in g_main_context_iteration (context=context@entry=0x55d413982a80, may_block=may_block@entry=1) at ../../../glib/gmain.c:4282
#16 0x00007fd96235950d in g_application_run (application=application@entry=0x55d4139859b0 [EvApplication], argc=argc@entry=1, argv=argv@entry=0x7ffcca9793f8)
    at ../../../gio/gapplication.c:2712
#17 0x000055d410ca7012 in main (argc=1, argv=0x7ffcca9793f8) at ../src/main.c:42
```

Both are related to getting the client from the list model.

EDIT: I deliberately didn't remove any secrets or db yet

fortysixandtwo commented:
strange, I cannot reproduce these on my end
(I've not reset the db/secrets since a couple of hours when I first got it running)

I got criticals (only once, they disappeared after rerunning to enable `G_DEBUG=fatal-criticals`), but they are of a completely different kind (and I raised an issue in [libcmatrix](https://source.puri.sm/Librem5/libcmatrix/-/issues/40) about them):


EDIT: I should read more closely: You said this critical/crashed on **shutdown**.
I'll keep my eyes peeled for this one.

fortysixandtwo commented:
Isn't that crash on shutdown perhaps fixed by https://source.puri.sm/Librem5/libcmatrix/-/merge_requests/86 ?

agx commented:
> Isn't that crash on shutdown perhaps fixed by https://source.puri.sm/Librem5/libcmatrix/-/merge_requests/86 ?

That one fixes both the ones on startup and shutdown. For future MRs feel free to pull in all relevant patches into the branch pointed at by the subproject wrap.

agx commented:
All pending libcmatrix MRs for this are in as far as I can tell so you could switch the libcmatrix wrap back to the blessed repo's `main` branch.

fortysixandtwo commented:
v4.1: switch to the main branch of the blessed repo (Librem5/libcmatrix)

agx commented:
I've made some minor cleanups like adding the `FIXME:` and pushed this.

Thanks a lot for fixing this, this will make using eigenvalue a lot nicer.
